mlflow/types/utils.py

```diff
@@ -79,7 +79,90 @@ def clean_tensor_type(dtype: np.dtype):
     return dtype
 
 
-def _infer_datatype(data) -> DataType:
+def _infer_colspec_type(data: Any) -> Union[DataType, Array, Object]:
+    """
+    Infer an MLflow Colspec type from the dataset.
+
+    :param data: data to infer from.
+    :return: Object
+    """
+    dtype = _infer_datatype(data)
+
+    # Currently only input that gives None is nested list whose items are all empty e.g. [[], []],
+    # because flat empty list [] has special handlign logic in _infer_schema
+    if dtype is None:
+        raise MlflowException(
+            "A column of nested array type must include at least one non-empty array."
+        )
+
+    return dtype
+
+
+def _infer_datatype(data: Any) -> Union[DataType, Array, Object]:
+    if isinstance(data, dict):
+        properties = []
+        for k, v in data.items():
+            dtype = _infer_datatype(v)
+            if dtype is None:
+                raise MlflowException("Property value must not be an empty list.")
```

Comment:
```suggestion
                raise MlflowException("Dictionary values must not be an empty list.")
```
As this function is called when inferring schema, users might not know what property is.

Reply:
Hmm I used 'property' as a general json-related term, rather than our internal class. Do we expose term "Dictionary" anywhere else?

Reply:
btw constructor exceptions say "Properties" or "Object"

Reply:
Because in such case users are inferring schema from a dictionary item

Reply:
Ok, I don't have strong opinion, but probably we need consistency. Apparently some message says "json" / "properties" still. I can clean them up as follow-up:)

Reply:
Emm we're not using json here, why should we mention json?

Reply:
I mean the opposite, like some messages say json / property but for user it's still dictionary right?